src/b-splines/cubic.jl

```diff
@@ -0,0 +1,248 @@
+immutable Cubic{BC<:Flag} <: Degree{3} end
+Cubic{BC<:Flag}(::BC) = Cubic{BC}()
+
+"""
+Assuming uniform knots with spacing 1, the `i`th piece of cubic spline
+implemented here is defined as follows.
+
+    y_i(x) = ϕm p(x-i) + ϕ q(x-i) + ϕp q(1- (x-i)) + ϕpp p(1 - (x-i))
+
+where
+
+    p(☆) = 1/6 * (1-☆)^3
+    q(☆) = 2/3 - ☆^2 + 1/2 ☆^3
+
+and the values `ϕX` for `X ⋹ {m, _, p, pp}` are the pre-filtered coefficients.
+
+For future reference, this expands out to the following polynomial:
+
+    y_i(x) = 1/6 cm (1+i-x)^3 + c (2/3 - (x-i)^2 + 1/2 (x-i)^3) +
+             cp (2/3 - (1+i-x)^2 + 1/2 (1+i-x)^3) + 1/6 cpp (x-i)^3
+
+When we derive boundary conditions we will use derivatives `y_0'(x)` and
+`y_0''(x)`
+"""
+Cubic
+
+function define_indices_d{BC}(::Type{BSpline{Cubic{BC}}}, d, pad)
+    symix, symixm, symixp = symbol("ix_",d), symbol("ixm_",d), symbol("ixp_",d)
+    symixpp, symx, symfx = symbol("ixpp_",d), symbol("x_",d), symbol("fx_",d)
+    quote
+        # ensure that all of ix_d, ixm_d, ixp_d, and ixpp_d are in-bounds no
+        # matter the value of pad
+        $symix = clamp(round(Int, real($symx)), 2-$pad, size(itp,$d)+$pad-2)
+        $symfx = $symx - $symix
+        $symix += $pad # padding for oob coefficient
+        $symixm = $symix - 1
+        $symixp = $symix + 1
+        $symixpp = $symixp + 1
+    end
+end
+
+# TODO: not sure if I am doing any of this Periodic stuff correctly
+function define_indices_d(::Type{BSpline{Cubic{Periodic}}}, d, pad)
+    symix, symixm, symixp = symbol("ix_",d), symbol("ixm_",d), symbol("ixp_",d)
+    symixpp, symx, symfx = symbol("ixpp_",d), symbol("x_",d), symbol("fx_",d)
+    quote
+        $symix = clamp(round(Int, real($symx)), 1, size(itp,$d))
+        $symfx = $symx - $symix
+        $symixm = mod1($symix - 1, size(itp,$d))
+        $symixp = mod1($symix + 1, size(itp,$d))
+        $symixpp = mod1($symix + 2, size(itp,$d))
+    end
+end
+
+padding{BC<:Flag}(::Type{BSpline{Cubic{BC}}}) = Val{1}()
+padding(::Type{BSpline{Cubic{Periodic}}}) = Val{0}()
```

Comment:
This looks good to me :+1:

The idea with padding is to give room for the "ghost cells" needed to implement the boundary conditions. For quadratic splines, where the i:th interpolation function is based on the coefficients at i-1, i and i+1, and yields values from i-1/2 to i+1/2. Thus, one needs to pad with 1 at the bottom (since at i==1 we also need the 0:th coefficient) and 1 at the top (since at i==N we also need the N+1:th).

Similarly, for cubic splines we use the i:th function to interpolate between i and i+1 (note that this is *not* centered on i, like the quadratic case), and uses four points from i-1 to i+2. Thus, when i==1 we need the 0:th coefficient, and at i==N we can use the N-1:th function and get by with padding only to N+1 here, i.e. a padding of 1 on each end here too.

It strikes me as I'm writing this that this only applies for `OnGrid` cubic interpolation; for `OnCell`, we're either continuing the interpolant on [1,2) for an extra half-cell below (this seems to be what we're doing now) or we need to add another padding coefficient to be able to interpolate the half-cell with the interpolant valid on [0,1). I think it's fine to do it this way, but we might want to document it explicitly.